### ISSUE TITLE
flake: TestCrossNamespaceListWatch: Unable to decode an event from the watch stream

### ISSUE THREAD

**Issue by sjenning** [kubernetes/kubernetes#46446]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46371/pull-kubernetes-unit/32665/

Cause of failure is
```
streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: Object 'Kind' is missing in '{"metadata":{},"status":"Failure","message":"too old resource version: 8420 (8421)","reason":"Gone","code":410}'
```

Full log:
```
=== RUN   TestCrossNamespaceListWatch
I0525 09:16:23.530175   16520 naming_controller.go:295] Shutting down NamingConditionController
I0525 09:16:23.530208   16520 customresource_discovery_controller.go:163] Shutting down DiscoveryController
I0525 09:16:23.530233   16520 crd_finalizer.go:261] Shutting down CRDFinalizer
I0525 09:16:23.530289   16520 serve.go:134] Stopped listening on 127.0.0.1:39195
2017-05-25T09:16:23.407225173-07:00 AUDIT: id="0d3b381e-a3db-452c-887f-eeefed0f8c5a" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions?resourceVersion=8402&timeoutSeconds=387&watch=true" response="200"
W0525 09:16:23.625212   16520 authentication.go:50] Authentication is disabled
I0525 09:16:23.628000   16520 serve.go:85] Serving securely on 127.0.0.1:43271
I0525 09:16:23.628181   16520 crd_finalizer.go:249] Starting CRDFinalizer
I0525 09:16:23.628213   16520 customresource_discovery_controller.go:152] Starting DiscoveryController
I0525 09:16:23.628228   16520 naming_controller.go:284] Starting NamingConditionController
2017-05-25T09:16:23.634800027-07:00 AUDIT: id="1a6e4b28-5fd8-4153-8499-19e1e6430232" ip="127.0.0.1" method="list" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions?resourceVersion=0" response="200"
2017-05-25T09:16:23.636170956-07:00 AUDIT: id="8db3124b-0371-4064-a1dd-b2af48962395" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions?resourceVersion=8413&timeoutSeconds=361&watch=true" response="<deferred>"
2017-05-25T09:16:23.658618061-07:00 AUDIT: id="2f470e31-b66b-4932-8fd2-11480be2bdb6" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/healthz" response="200"
2017-05-25T09:16:23.659536141-07:00 AUDIT: id="c29896c5-1192-458f-9c86-4f1b6bf67fd1" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1" response="200"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"apiextensions.k8s.io/v1alpha1","resources":[{"name":"customresourcedefinitions","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["create","delete","deletecollection","get","list","patch","update","watch"]},{"name":"customresourcedefinitions/status","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["update"]}]}
2017-05-25T09:16:23.660440595-07:00 AUDIT: id="51d98f8a-0bd9-4e18-b9bf-fd8fac6118d6" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions" response="201"
2017-05-25T09:16:23.663298601-07:00 AUDIT: id="c0037908-b4e2-4220-8754-f2d5f4b65abf" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-25T09:16:23.694378381-07:00 AUDIT: id="6ce6cb6a-cec6-4eb1-bd25-c540bc20297d" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-25T09:16:23.724342634-07:00 AUDIT: id="77b91864-99d1-4b49-b201-2b4866f04b6b" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-25T09:16:23.729122625-07:00 AUDIT: id="c275c197-8a5c-4a5e-9fa9-ef37dd2bff60" ip="127.0.0.1" method="update" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions/noxus.mygroup.example.com/status" response="200"
2017-05-25T09:16:23.732372793-07:00 AUDIT: id="3666a3ff-df4e-4eaf-aecd-d2eaaff0f3ff" ip="127.0.0.1" method="update" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions/noxus.mygroup.example.com/status" response="409"
E0525 09:16:23.733139   16520 naming_controller.go:316] noxus.mygroup.example.com failed with: the server reported a conflict (put customresourcedefinitions.apiextensions.k8s.io noxus.mygroup.example.com)
2017-05-25T09:16:23.754512366-07:00 AUDIT: id="e84a520b-debb-41bc-9b4c-32299fc77ac9" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="200"
2017-05-25T09:16:23.75567665-07:00 AUDIT: id="d077799d-1428-4c0a-aa3a-058b2de5ea95" ip="127.0.0.1" method="list" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus" response="200"
2017-05-25T09:16:23.760866494-07:00 AUDIT: id="ae6a645d-a7f4-4724-a026-ce1d01ac3428" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus?resourceVersion=8420&watch=true" response="<deferred>"
2017-05-25T09:16:23.762105972-07:00 AUDIT: id="c0605d78-9421-4648-8106-784fc848b2d4" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-1" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-1/noxus" response="201"
2017-05-25T09:16:23.7650626-07:00 AUDIT: id="eaee5741-5030-468c-b512-274bd9916f5b" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-1" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-1/noxus?resourceVersion=8420&watch=true" response="<deferred>"
2017-05-25T09:16:23.7650626-07:00 AUDIT: id="eaee5741-5030-468c-b512-274bd9916f5b" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-1" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-1/noxus?resourceVersion=8420&watch=true" response="200"
E0525 09:16:23.765876   16520 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: Object 'Kind' is missing in '{"metadata":{},"status":"Failure","message":"too old resource version: 8420 (8421)","reason":"Gone","code":410}'
2017-05-25T09:16:23.765878749-07:00 AUDIT: id="4bd54474-8c25-495b-aea2-1af413fcd1c3" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-2" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-2/noxus" response="201"
2017-05-25T09:16:23.767748183-07:00 AUDIT: id="895343c0-0f3a-475f-8c14-0d01b820c2af" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-2" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-2/noxus?resourceVersion=8420&watch=true" response="<deferred>"
2017-05-25T09:16:23.767748183-07:00 AUDIT: id="895343c0-0f3a-475f-8c14-0d01b820c2af" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="namespace-2" uri="/apis/mygroup.example.com/v1alpha1/namespaces/namespace-2/noxus?resourceVersion=8420&watch=true" response="200"
E0525 09:16:23.768395   16520 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: Object 'Kind' is missing in '{"metadata":{},"status":"Failure","message":"too old resource version: 8420 (8421)","reason":"Gone","code":410}'
2017-05-25T09:16:23.768447796-07:00 AUDIT: id="d2b12249-f743-4ef5-b12f-efadece43220" ip="127.0.0.1" method="list" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus" response="200"
--- FAIL: TestCrossNamespaceListWatch (0.24s)
	basic_test.go:501: expected ADDED, got 
=== RUN   TestMultipleResourceInstances
```

**Comment by ncdc** [comment]
FYI @deads2k @kubernetes/sig-api-machinery-misc

**Comment by yujuhong** [comment]
Another one: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/44061/pull-kubernetes-unit/32758/

**Comment by NickrenREN** [comment]
This one also hit: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46484/pull-kubernetes-unit/32866/

**Comment by deads2k** [comment]
I'm looking at it.  I'm going to try to see what's up with the watch cache first, but failing that I'll make it tolerant of having to relist I suppose.

**Comment by rmmh** [comment]
Here's a cluster to track them https://storage.googleapis.com/k8s-gubernator/triage/index.html?text=expected%20ADDED#e4d5a498d9c644be6c28

**Comment by deads2k** [comment]
I think I found it and opened https://github.com/kubernetes/kubernetes/issues/46536

**Comment by liggitt** [comment]
I can see how #46536 would avoid the "too old" error, but in cases where the "too old" error is actually supposed to be getting returned, don't we have a separate problem with the watch event error not being able to be decoded?

**Comment by liggitt** [comment]
still flaking
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-test-go/6169/
ran on b9e8d2aee6d59330306cb458c1241f5e2578c40b (two commits after #46536 merged)
